src/snowflake/snowpark/_internal/ast/builder.py

```diff
@@ -0,0 +1,159 @@
+        """
+        if dependency not in self.__dependencies:
+            self.__dependencies.append(dependency)
+        if self not in dependency.__required_by:
+            dependency.__required_by.append(self)
+
+    @staticmethod
+    def _closure(root: "AstBuilder") -> List["AstBuilder"]:
+        """
+        Performs a post-order traversal of the AST to compute the transitive closure of dependencies.
+        This ensures that all dependencies are included in the final AST request.
+
+        Args:
+            root: The root AstBuilder instance from which to compute the closure.
+        """
+        q: deque[AstBuilder] = deque()
+        q.append(root)
+
+        visited = set()
+        closure = []
+
+        while q:
+            ast_builder = q.pop()
+            closure.append(ast_builder)
+            visited.add(ast_builder)
+            for dep in ast_builder.__dependencies:
+                if dep not in visited:
+                    q.append(dep)
+        closure.reverse()
+
+        return closure
+
+    def to_request(self, assign_first_request_ids: bool = True) -> SerializedBatch:
+        """
+        Builds and returns a new AST request with transitive closure of all dependencies.
+        """
+        request = proto.Request()
+
+        # Generate a new unique ID and convert to bytes for the request message field.
+        current_request_id: uuid.UUID = AstBuilder.__generate_request_id()
+        request.id = current_request_id.bytes
+
+        # Initialize the request with the current client version, Python version, and AST version.
+        (major, minor, patch) = VERSION
+        request.client_version.major = major
+        request.client_version.minor = minor
+        request.client_version.patch = patch
+
+        (major, minor, micro, releaselevel, serial) = sys.version_info
+        request.client_language.python_language.version.major = major
+        request.client_language.python_language.version.minor = minor
+        request.client_language.python_language.version.patch = micro
+        request.client_language.python_language.version.label = releaselevel
+
+        request.client_ast_version = CLIENT_AST_VERSION
+
+        # Get the transitive closure of the current AST builder and its dependencies.
+        # Add all statements in the closure to the request body with unique IDs.
+        closure = AstBuilder._closure(self)
+        for ast_handle in closure:
+            if assign_first_request_ids and not ast_handle._bind.first_request_id:
+                # Assign the first request IDs for bind statements if not already set.
+                ast_handle._bind.first_request_id = current_request_id.bytes
+
+            # Add the bind and eval statements to the request body.
+            request.body.append(ast_handle._bind)
+            if ast_handle.__eval is not None:
+                request.body.append(ast_handle.__eval)
+
+        if self.__eval is None:
```

Comment:
this makes the function non-const and have side effects. Could we maybe not overwrite internal `self.__eval`? I actually think you don't need to even add __eval to the class. Could be fine to add here (e.g., introduce param `eval=False` and then explicitly call with `eval=True` in the source code). Wdyt?

Reply:
I don't think `__eval` is necessary either in this class either since we are moving to globally unique Bind IDs, unless we'd like to include previously generated Eval statements in future batches as a hint to the server. Otherwise I agree we should move to introduce an `eval=False` param to this method.

Reply:
Ok, then let's remove the __eval here (or add with eval=False/True). Good to have stateless/not self modifying function here.